src/transport_ofi.c

```diff
@@ -1329,6 +1333,98 @@ int allocate_fabric_resources(struct fabric_info *info)
     return ret;
 }
 
+#ifdef USE_HWLOC
+struct fi_info *assign_nic_with_hwloc(struct fi_info *fabric, struct fi_info **provs, size_t num_nics) {
+    int ret = 0;
+    hwloc_bitmap_t bindset = hwloc_bitmap_alloc();
+
+    ret = hwloc_get_proc_last_cpu_location(shmem_topology, getpid(), bindset, HWLOC_CPUBIND_PROCESS);
+    if (ret < 0) {
+        RAISE_ERROR_MSG("hwloc_get_proc_last_cpu_location failed (%s)\n", strerror(errno));
+    }
+
+    // Identify which provider entries correspond to NICs with an affinity to the calling process
+    struct fi_info *close_provs = NULL;
+    struct fi_info *last_added = NULL;
+    size_t num_close_nics = 0;
+    for (size_t i = 0; i < num_nics; i++) {
+        struct fi_info *cur_prov = provs[i];
+        if (cur_prov->nic->bus_attr->bus_type != FI_BUS_PCI) continue;
+
+        struct fi_pci_attr pci = cur_prov->nic->bus_attr->attr.pci;
+        hwloc_obj_t io_device = hwloc_get_pcidev_by_busid(shmem_topology, pci.domain_id, pci.bus_id, pci.device_id, pci.function_id);
+        if (!io_device) RAISE_ERROR_MSG("hwloc_get_pcidev_by_busid failed\n");
+        hwloc_obj_t first_non_io = hwloc_get_non_io_ancestor_obj(shmem_topology, io_device);
+        if (!first_non_io) RAISE_ERROR_MSG("hwloc_get_non_io_ancestor_obj failed\n");
```

Comment:
Is it possible to fallback to round-robin if one of these routines return a NULL object?  (If so, I'd think a warning message may be appropriate).